src/Facebook/Monad.hs

```diff
@@ -112,38 +124,72 @@ data FbTier
   -> H.Manager -- ^ Connection manager (see 'H.withManager').
   -> FacebookT Auth m a
   -> m a
-runFacebookT creds manager (F act) =
-  runReaderT act (FbData creds manager Production)
+runFacebookT creds apiVersion manager (F act) =
+  runReaderT act (FbData ( Just creds ) manager Production apiVersion)
+
+
+addAppSecretProof :: Credentials
+                  -> Maybe (AccessToken anykind)
+                  -> HT.SimpleQuery
+                  -> HT.SimpleQuery
+addAppSecretProof ( Credentials _ _ _ False ) _ query = query
+addAppSecretProof creds mtoken query = makeAppSecretProof creds mtoken <> query
+
+-- | Make an appsecret_proof in case the given credentials access token is a
+-- user access token.
+-- See: https://developers.facebook.com/docs/graph-api/securing-requests/#appsecret_proof
+makeAppSecretProof
+  :: Credentials -- ^ App credentials
+  -> Maybe ( AccessToken anyKind ) -- ^
+  -> HT.SimpleQuery
+makeAppSecretProof creds (Just ( UserAccessToken _ accessToken _ ))
+  = [( TE.encodeUtf8 "appsecret_proof", proof)]
+
+  where
+    key :: MacKey ctx SHA256
+    key   = MacKey $ appSecretBS creds
+    proof = Cereal.encode $ hmac' key ( TE.encodeUtf8 accessToken )
+makeAppSecretProof  _ _ = []
+
 
 -- | Run a computation in the 'FacebookT' monad without
 -- credentials.
-runNoAuthFacebookT :: H.Manager -> FacebookT NoAuth m a -> m a
-runNoAuthFacebookT manager (F act) =
-  let creds = error "runNoAuthFacebookT: never here, serious bug"
-  in runReaderT act (FbData creds manager Production)
+runNoAuthFacebookT
+  :: ApiVersion -- ^ Api version
+  -> H.Manager -- ^ Connection manager (see 'H.withManager').
+  -> FacebookT NoAuth m a -> m a
+runNoAuthFacebookT apiVersion manager (F act) =
+  runReaderT act (FbData Nothing manager Production apiVersion)
+
 
 -- | Same as 'runFacebookT', but uses Facebook's beta tier (see
 -- <https://developers.facebook.com/support/beta-tier/>).
-beta_runFacebookT :: Credentials -> H.Manager -> FacebookT Auth m a -> m a
-beta_runFacebookT creds manager (F act) =
-  runReaderT act (FbData creds manager Beta)
+beta_runFacebookT :: Credentials -> ApiVersion -> H.Manager -> FacebookT Auth m a -> m a
+beta_runFacebookT creds apiVersion manager (F act) =
+  runReaderT act (FbData (Just creds) manager Beta apiVersion)
 
 -- | Same as 'runNoAuthFacebookT', but uses Facebook's beta tier
 -- (see <https://developers.facebook.com/support/beta-tier/>).
-beta_runNoAuthFacebookT :: H.Manager -> FacebookT NoAuth m a -> m a
-beta_runNoAuthFacebookT manager (F act) =
-  let creds = error "beta_runNoAuthFacebookT: never here, serious bug"
-  in runReaderT act (FbData creds manager Beta)
+beta_runNoAuthFacebookT :: ApiVersion -> H.Manager -> FacebookT NoAuth m a -> m a
+beta_runNoAuthFacebookT apiVersion manager (F act) =
+  runReaderT act (FbData Nothing manager Beta apiVersion)
 
 -- | Get the user's credentials.
 getCreds
   :: Monad m
-  => FacebookT Auth m Credentials
+  => FacebookT anyAuth m (Maybe Credentials)
```

Comment:
Can we not change the type signature of this ? Instead if we get `Nothing` value - let's throw an exception: `FbLibraryException`. This will avoid partial `Just` pattern matching we have currently in the code.

Reply:
@psibi, If we do that, then `fbreq` type would need to change to have the auth parameter to be defined as `Auth`, and then we would need to duplicate `fbreq` to support the `NoAuth` calls.

Reply:
@damianfral You can have two methods: 
```
getCreds :: FacebookT anyAuth m Credentials
getMCreds :: FacebookT anyAuth m (Maybe Credentials)
```

Reply:
@psibi, done! :)